# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: **Gustavo Roger Bravo Esquivias**  <br>
Url: https://github.com/gussbrav/SEMINARIO<br>
Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:

# **Combinar cifras y operaciones**

• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.

• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)

• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4

Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo ?

• Nota: Es posible usar la función de python “eval” para evaluar una expresión, ejemplo:
print(eval('4+2-6/3*1'))
resultado=4


(*) La respuesta es obligatoria





                                        

## Detalles a tener en cuenta
Para resolver este problema, primero se generan todas las formas posibles de ordenar los números del 1 al 9, junto con todas las combinaciones de los operadores matemáticos básicos (+, -, *, /). Luego, para cada posible disposición de cifras, se prueba con cada secuencia de operadores, formando expresiones matemáticas válidas. Cada expresión generada se evalúa y, si su resultado es un número entero, se almacena en un conjunto para su posterior análisis.

Una vez obtenidos los resultados enteros, se identifican el valor mínimo y el valor máximo dentro del conjunto. Luego, se verifica si todos los números enteros dentro de ese rango están presentes sin interrupciones. Finalmente, se presentan los resultados mostrando el número mínimo, el máximo y si la secuencia de valores es continua.




In [2]:
import itertools
import math

# Generadores originales
permutaciones_numericas = itertools.permutations(map(str, range(1, 10)))
permutaciones_simbolicas = list(itertools.permutations(['+', '-', '*', '/']))

expresiones_validas = {}
resultados_obtenidos = set()

def fusionar_componentes(numeros, simbolos):
    expresion = ''
    for i in range(len(simbolos)):
        expresion += numeros[i] + simbolos[i]
    expresion += numeros[-1]
    return expresion

# Evaluar expresiones
for nums in permutaciones_numericas:
    for ops in permutaciones_simbolicas:
        try:
            expr = fusionar_componentes(nums, ops)
            resultado = eval(expr)
            if resultado == int(resultado):
                resultados_obtenidos.add(int(resultado))
                expresiones_validas[expr] = resultado
        except ZeroDivisionError:
            continue

# Resultados
if resultados_obtenidos:
    valor_minimo = min(resultados_obtenidos)
    valor_maximo = max(resultados_obtenidos)
    todos_valores_presentes = all(x in resultados_obtenidos for x in range(valor_minimo, valor_maximo + 1))
else:
    valor_minimo = valor_maximo = None
    todos_valores_presentes = False

print(f"Permutaciones numéricas: {math.factorial(9)}, Permutaciones simbólicas: {len(permutaciones_simbolicas)}, Expresiones generadas: {len(expresiones_validas)}")
print(f"Valor mínimo: {valor_minimo}, Valor máximo: {valor_maximo}, Todos presentes: {todos_valores_presentes}")


Permutaciones numéricas: 362880, Permutaciones simbólicas: 24, Expresiones generadas: 90000
Valor mínimo: -69, Valor máximo: 77, Todos presentes: True


## (*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

Para establecer la cantidad total de combinaciones posibles sin tomar en cuenta restricciones, se procede con el siguiente análisis:

* **Variaciones de los dígitos**  
  Consideramos 9 dígitos distintos que van del 1 al 9.  
  El número total de variaciones posibles es calculado mediante 9 factorial (9!), obteniendo así 362,880 variaciones.

* **Variaciones de los operadores matemáticos**  
  Disponemos de 4 operadores distintos (+, -, *, /).  
  La cantidad total de variaciones para estos operadores es de 4 factorial (4!), resultando en 24.

* **Total combinado de posibilidades**  
  Al multiplicar cada una de las variaciones numéricas por las variaciones de operadores, se obtiene un total global de combinaciones: 9! x 4! = 362,880 x 24 = **8,718,720 combinaciones totales**.

## ¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.

Cuando se consideran restricciones específicas, se obtienen los siguientes resultados:

* **Variaciones numéricas**  
  Los mismos 9 dígitos del 1 al 9 generan nuevamente 362,880 variaciones posibles (9!).

* **Variaciones de operadores**  
  Se mantienen los 4 operadores matemáticos básicos (+, -, *, /), con un total de 24 variaciones posibles (4!).

* **Combinación de dígitos y operadores**  
  Cada una de las variaciones numéricas se combina con las 24 configuraciones diferentes de operadores.

Sin embargo, al evaluar estas combinaciones, no todas cumplen con el requisito de generar resultados enteros válidos. Al realizar una filtración adecuada considerando esta condición, se obtiene un número aproximado de **80,000 combinaciones válidas**.


Modelo para el espacio de soluciones<br>
# (*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)

Con respecto a la evaluación de las estructuras de datos utilizadas en el código
se tiene que abordar este problema de manera eficiente, es fundamental elegir estructuras de datos que optimicen tanto la generación de combinaciones como la evaluación y almacenamiento de resultados. A continuación, se analizan las estructuras empleadas en el código:

1. Listas (list) para almacenar permutaciones de cifras y operadores
Las listas resultan adecuadas para manejar las permutaciones generadas con itertools.permutations, ya que permiten un acceso rápido mediante índices y son sencillas de manipular.

Ventajas:
Permiten acceder a los elementos de forma directa mediante índices.
Son fáciles de construir y entender.
Facilitan la iteración eficiente sobre los elementos.

Desventajas:
No garantizan que los elementos sean únicos.
En estructuras grandes, las operaciones de búsqueda y eliminación pueden volverse costosas en términos de rendimiento.

2. Conjuntos (set) para almacenar los resultados enteros
El uso de conjuntos es una opción eficiente cuando se necesita garantizar la unicidad de los elementos almacenados. Dado que los conjuntos no permiten valores duplicados, son ideales para registrar los resultados de manera óptima.

Ventajas:
Aseguran que no haya duplicados en los datos almacenados.
Las operaciones de inserción, eliminación y búsqueda tienen una complejidad promedio de O(1), lo que mejora el rendimiento en comparación con listas en estos casos.


Según el modelo para el espacio de soluciones<br>
# (*)¿Cual es la función objetivo?

En este problema, el objetivo ta principal es analizar todas las expresiones posibles que se pueden formar al intercalar las permutaciones de las cifras con los operadores. Luego, se deben evaluar estas expresiones y determinar cuántas de ellas generan un resultado que sea un número entero.



# (*)¿Es un problema de maximización o minimización?

Este problema no es exclusivamente de maximización o minimización, sino un problema de optimización que abarca ambos aspectos para definir un rango de soluciones posibles.

# Diseña un algoritmo para resolver el problema por fuerza bruta

Respuesta: <br>
El algoritmo diseñado se realizo utilizando la fuerza bruta.

# Calcula la complejidad del algoritmo por fuerza bruta

Respuesta: <br>

In [1]:
# Importamos las librerías necesarias
import itertools as it
import numpy as np


# Generación de combinaciones y permutaciones


# Generamos todas las combinaciones posibles de 5 números únicos entre 1 y 9
combinaciones_numeros = list(it.combinations(range(1, 10), 5))
# Generamos las permutaciones de cada combinación
permutaciones_numeros = list(it.chain.from_iterable(it.permutations(c) for c in combinaciones_numeros))

# Generamos todas las permutaciones posibles de los operadores '+', '-', '*', '/'
permutaciones_operadores = list(it.permutations(['+', '-', '*', '/']))


# Función para construir una ecuación a partir de
# una permutación de números y operadores


def construir_ecuacion(numeros, operadores):
    """
    Recibe una tupla de números y una tupla de operadores.
    Retorna un string representando la ecuación.
    """
    return ''.join(f"{num}{op}" for num, op in zip(numeros, operadores)) + str(numeros[-1])


# Generar todas las ecuaciones y filtrar soluciones ####


def encontrar_ecuaciones_con_resultado(numeros_permutados, operadores_permutados, objetivo):
    """
    Genera todas las ecuaciones posibles y filtra aquellas cuyo resultado sea igual al número objetivo.
    """
    soluciones = []
    for numeros in numeros_permutados:
        for operadores in operadores_permutados:
            ecuacion = construir_ecuacion(numeros, operadores)
            try:
                if eval(ecuacion) == objetivo:
                    soluciones.append(ecuacion)
            except ZeroDivisionError:
                continue  # Ignorar divisiones por cero
    return soluciones


#Función principal para ejecutar el cálculo ####


def resolver_por_fuerza_bruta(objetivo):
    """
    Encuentra todas las ecuaciones que resultan en el número objetivo.
    """
    soluciones = encontrar_ecuaciones_con_resultado(permutaciones_numeros, permutaciones_operadores, objetivo)
    if soluciones:
        print(f"Las posibles soluciones para {objetivo} son:")
        for sol in soluciones:
            print(sol)
    else:
        print(f"No existe solución para el número: {objetivo}")


In [5]:
resolver_por_fuerza_bruta(77)

Las posibles soluciones para 77 son:
7/1-2+8*9
7/1-2+9*8
7/1+8*9-2
7/1+9*8-2
7-2/1+8*9
7-2/1+9*8
7-2+8/1*9
7-2+8*9/1
7-2+9/1*8
7-2+9*8/1
7+8/1*9-2
7+8*9/1-2
7+8*9-2/1
7+9/1*8-2
7+9*8/1-2
7+9*8-2/1
8/1*9-2+7
8/1*9+7-2
8*9/1-2+7
8*9/1+7-2
8*9-2/1+7
8*9-2+7/1
8*9+7/1-2
8*9+7-2/1
9/1*8-2+7
9/1*8+7-2
9*8/1-2+7
9*8/1+7-2
9*8-2/1+7
9*8-2+7/1
9*8+7/1-2
9*8+7-2/1
7-4/2+8*9
7-4/2+9*8
7+8*9-4/2
7+9*8-4/2
8*9-4/2+7
8*9+7-4/2
9*8-4/2+7
9*8+7-4/2
7-6/3+8*9
7-6/3+9*8
7+8*9-6/3
7+9*8-6/3
8*9-6/3+7
8*9+7-6/3
9*8-6/3+7
9*8+7-6/3


## Calcula la Complejidad total del algoritmo por fuerza bruta:
Podemos afirma que la generación de permutaciones de números y operadores es la parte más costosa en términos de tiempo de ejecución.

La complejidad del algoritmo por fuerza fruta sería factorial: $$ O(n!) $$
Por lo tanto **O(8718720)**, se puede simplificar a **O(9!×4!)**.



# (*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

## Algoritmo Genético
Una alternativa para mejorar la complejidad del algoritmo de fuerza bruta es utilizar un algoritmo genético.

El algoritmos genéticos es una técnica de optimización inspirada en la evolución natural, donde una población de soluciones evoluciona mediante procesos como selección, cruce (crossover) y mutación.

## Porque mejora el algoritmo de fuerza bruta:

Reduce el espacio de búsqueda seleccionando combinaciones prometedoras en cada generación.

Evita evaluar todas las permutaciones posibles, lo que ahorra tiempo de ejecución.

Encuentra soluciones cercanas al óptimo rápidamente, sin necesidad de revisar exhaustivamente todas las opciones.

Evita quedarse atrapado en soluciones subóptimas, gracias a los operadores de selección, cruce y mutación.

In [9]:
import itertools as it
import random

# Definir los números y operadores disponibles
numeros_disponibles = list(range(1, 10))  # Dígitos del 1 al 9
operadores_disponibles = ['+', '-', '*', '/']

# Función para generar una ecuación a partir de una combinación de números y operadores
def construir_ecuacion(numeros, operadores):
    return ''.join(f"{num}{op}" for num, op in zip(numeros, operadores)) + str(numeros[-1])

# Función de evaluación: retorna el valor de la expresión si es un número entero, sino None
def evaluar_ecuacion(ecuacion):
    try:
        resultado = eval(ecuacion)
        return int(resultado) if resultado == int(resultado) else None
    except ZeroDivisionError:
        return None

# Generación de la población inicial con mayor diversidad
def generar_poblacion(tamano):
    poblacion = []
    for _ in range(tamano):
        numeros = random.sample(numeros_disponibles, 5)  # Seleccionar 5 números únicos
        operadores = random.choices(operadores_disponibles, k=4)  # Elegir 4 operadores
        poblacion.append((numeros, operadores))
    return poblacion

# Función de fitness: favorece resultados enteros y prioriza valores cercanos a 0
def calcular_fitness(individuo):
    numeros, operadores = individuo
    ecuacion = construir_ecuacion(numeros, operadores)
    resultado = evaluar_ecuacion(ecuacion)

    if resultado is None:
        return -1000  # Penaliza expresiones no válidas

    return -abs(resultado)  # Favorece valores cercanos a 0

# Operador de selección (torneo con elitismo)
def seleccion(poblacion, fitness_scores):
    mejores = sorted(zip(poblacion, fitness_scores), key=lambda x: x[1], reverse=True)
    return mejores[0][0] if random.random() < 0.1 else random.choice(mejores[:5])[0]  # Elitismo + torneo

# Operador de cruce (crossover mejorado)
def cruce(padre1, padre2):
    punto_corte = random.randint(1, 3)  # Punto de cruce entre operadores
    nuevo_operadores = padre1[1][:punto_corte] + padre2[1][punto_corte:]
    nuevo_numeros = list(set(padre1[0][:3] + padre2[0][2:]))[:5]  # Mezcla números únicos
    return nuevo_numeros, nuevo_operadores

# Operador de mutación corregido
def mutacion(individuo):
    if random.random() < 0.3:  # Mayor probabilidad de mutación
        idx_op = random.randint(0, len(individuo[1]) - 1)  # Índice seguro para operadores
        idx_num = random.randint(0, len(individuo[0]) - 1)  # Índice seguro para números

        # Mutar operador
        individuo[1][idx_op] = random.choice(operadores_disponibles)

        # Mutar número asegurando que no se repita en el individuo
        nuevo_numero = random.choice(numeros_disponibles)
        while nuevo_numero in individuo[0]:
            nuevo_numero = random.choice(numeros_disponibles)
        individuo[0][idx_num] = nuevo_numero

    return individuo

# Algoritmo Genético Principal
def algoritmo_genetico(iteraciones=300, tamano_poblacion=100):
    poblacion = generar_poblacion(tamano_poblacion)

    for _ in range(iteraciones):
        # Evaluar la población
        fitness_scores = [calcular_fitness(ind) for ind in poblacion]

        # Seleccionar los mejores individuos para la siguiente generación
        nueva_poblacion = []
        for _ in range(tamano_poblacion // 2):
            padre1 = seleccion(poblacion, fitness_scores)
            padre2 = seleccion(poblacion, fitness_scores)
            hijo1 = cruce(padre1, padre2)
            hijo2 = cruce(padre2, padre1)
            nueva_poblacion.extend([mutacion(hijo1), mutacion(hijo2)])

        poblacion = nueva_poblacion  # Actualizar la población

    # Encontrar la mejor ecuación con resultado 0
    mejor_expresion, mejor_resultado = None, None
    valor_minimo, valor_maximo = float('inf'), float('-inf')
    expresion_minima, expresion_maxima = None, None

    for individuo in poblacion:
        ecuacion = construir_ecuacion(individuo[0], individuo[1])
        resultado = evaluar_ecuacion(ecuacion)

        if resultado is not None:
            # Buscar mejor expresión con resultado 0
            if resultado == 0 and (mejor_resultado is None or abs(resultado) < abs(mejor_resultado)):
                mejor_expresion = ecuacion
                mejor_resultado = resultado

            # Actualizar valores mínimos y máximos
            if resultado < valor_minimo:
                valor_minimo = resultado
                expresion_minima = ecuacion
            if resultado > valor_maximo:
                valor_maximo = resultado
                expresion_maxima = ecuacion

    return mejor_expresion, mejor_resultado, valor_minimo, expresion_minima, valor_maximo, expresion_maxima

# Ejecutar el algoritmo
mejor_expresion, mejor_resultado, valor_minimo, expresion_minima, valor_maximo, expresion_maxima = algoritmo_genetico()

# Imprimir resultados en el formato esperado
print(f"Mejor expresión: {mejor_expresion}")
print(f"Resultado: {mejor_resultado}")
print(f"Valor mínimo: {valor_minimo} con expresión: {expresion_minima}")
print(f"Valor máximo: {valor_maximo} con expresión: {expresion_maxima}")
print((valor_minimo, valor_maximo, expresion_minima, expresion_maxima))


Mejor expresión: 8+2+4-7-7
Resultado: 0
Valor mínimo: -38 con expresión: 8+2+1-7*7
Valor máximo: 38 con expresión: 8*6+4-7-7
(-38, 38, '8+2+1-7*7', '8*6+4-7-7')


# (*)Calcula la complejidad del algoritmo

Respuesta

El algoritmo genético mejora la eficiencia en comparación con la fuerza bruta al reducir drásticamente el número de combinaciones evaluadas. Su complejidad depende de varios factores clave:

Generación de la población inicial:

Se generan P individuos en la población inicial, cada uno con combinaciones aleatorias de números y operadores.
La generación de cada individuo requiere seleccionar 5 números únicos (O(5)) y 4 operadores (O(4)).
Complejidad: O(P).
Evaluación de la función fitness:

En cada iteración, se evalúan todas las ecuaciones de la población (P individuos).
La evaluación de cada ecuación mediante eval() es O(1), ya que opera sobre una expresión matemática de tamaño fijo.
Complejidad por iteración: O(P).
Selección, cruce y mutación:

La selección de los mejores individuos mediante torneo (O(1)) se realiza P/2 veces.
El cruce genera dos nuevos individuos (O(1)).
La mutación reemplaza un operador o un número (O(1)).
Complejidad por iteración: O(P).
Número total de iteraciones (G):

El proceso evolutivo se repite durante G generaciones.
Complejidad total: O(P × G).

El algoritmo genético tiene una complejidad de O(P × G), donde P es el tamaño de la población y G el número de generaciones. En la práctica, P y G son significativamente menores que el número total de combinaciones posibles (9! × 4! ≈ 8.7 millones), lo que lo hace mucho más eficiente que la fuerza bruta (O(9! × 4!)). Esto permite encontrar soluciones óptimas en menos tiempo sin necesidad de evaluar todas las posibilidades.

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

In [13]:
import random
import pandas as pd

# Definir los números y operadores disponibles
numeros_disponibles = list(range(1, 10))  # Dígitos del 1 al 9
operadores_disponibles = ['+', '-', '*', '/']

# Función para generar una ecuación a partir de números y operadores
def construir_ecuacion(numeros, operadores):
    return ''.join(f"{num}{op}" for num, op in zip(numeros, operadores)) + str(numeros[-1])

# Función para evaluar una ecuación y devolver el resultado si es entero
def evaluar_ecuacion(ecuacion):
    try:
        resultado = eval(ecuacion)
        return int(resultado) if resultado == int(resultado) else None
    except ZeroDivisionError:
        return None

# Función para generar datos de entrada aleatorios
def generar_datos_entrada_aleatorios(cantidad_muestras=10):
    datos = []
    for _ in range(cantidad_muestras):
        # Seleccionar 5 números únicos aleatorios
        numeros = random.sample(numeros_disponibles, 5)
        # Seleccionar 4 operadores aleatorios
        operadores = random.choices(operadores_disponibles, k=4)
        # Construir la ecuación como string
        ecuacion = construir_ecuacion(numeros, operadores)
        # Evaluar la ecuación
        resultado = evaluar_ecuacion(ecuacion)
        datos.append((numeros, operadores, ecuacion, resultado))

    return datos

# Generamos el juego de datos de entrada aleatorios
datos_entrada_aleatorios = generar_datos_entrada_aleatorios(10)

# Convertimos los datos en un DataFrame para mejor visualización
df_datos_entrada = pd.DataFrame(datos_entrada_aleatorios, columns=["Números", "Operadores", "Ecuación Generada", "Resultado"])

# Mostrar los datos generados
print(df_datos_entrada)


           Números    Operadores Ecuación Generada  Resultado
0  [9, 4, 8, 1, 6]  [*, *, -, +]         9*4*8-1+6      293.0
1  [5, 6, 7, 3, 8]  [/, *, *, -]         5/6*7*3-8        NaN
2  [8, 9, 2, 7, 4]  [/, /, /, *]         8/9/2/7*4        NaN
3  [5, 8, 1, 7, 4]  [+, +, /, *]         5+8+1/7*4        NaN
4  [7, 8, 9, 3, 1]  [*, -, +, -]         7*8-9+3-1       49.0
5  [1, 5, 2, 7, 3]  [*, -, -, *]         1*5-2-7*3      -18.0
6  [9, 1, 8, 3, 7]  [-, -, -, /]         9-1-8-3/7        NaN
7  [6, 9, 8, 1, 7]  [-, -, /, *]         6-9-8/1*7      -59.0
8  [2, 5, 8, 6, 4]  [*, +, -, -]         2*5+8-6-4        8.0
9  [1, 2, 8, 7, 6]  [-, /, -, +]         1-2/8-7+6        NaN


Aplica el algoritmo al juego de datos generado

Respuesta

In [14]:
import random
import pandas as pd

# Definir los números y operadores disponibles
numeros_disponibles = list(range(1, 10))  # Dígitos del 1 al 9
operadores_disponibles = ['+', '-', '*', '/']

# Función para generar una ecuación a partir de números y operadores
def construir_ecuacion(numeros, operadores):
    return ''.join(f"{num}{op}" for num, op in zip(numeros, operadores)) + str(numeros[-1])

# Función para evaluar una ecuación y devolver el resultado si es entero
def evaluar_ecuacion(ecuacion):
    try:
        resultado = eval(ecuacion)
        return int(resultado) if resultado == int(resultado) else None
    except ZeroDivisionError:
        return None

# Función para generar datos de entrada aleatorios
def generar_datos_entrada_aleatorios(cantidad_muestras=10):
    datos = []
    for _ in range(cantidad_muestras):
        numeros = random.sample(numeros_disponibles, 5)
        operadores = random.choices(operadores_disponibles, k=4)
        ecuacion = construir_ecuacion(numeros, operadores)
        resultado = evaluar_ecuacion(ecuacion)
        datos.append((numeros, operadores, ecuacion, resultado))
    return datos

# Generación de la población inicial
def generar_poblacion(tamano):
    poblacion = []
    for _ in range(tamano):
        numeros = random.sample(numeros_disponibles, 5)
        operadores = random.choices(operadores_disponibles, k=4)
        poblacion.append((numeros, operadores))
    return poblacion

# Función de fitness
def calcular_fitness(individuo):
    numeros, operadores = individuo
    ecuacion = construir_ecuacion(numeros, operadores)
    resultado = evaluar_ecuacion(ecuacion)
    return -1000 if resultado is None else -abs(resultado)

# Selección por torneo con elitismo
def seleccion(poblacion, fitness_scores):
    mejores = sorted(zip(poblacion, fitness_scores), key=lambda x: x[1], reverse=True)
    return mejores[0][0] if random.random() < 0.1 else random.choice(mejores[:5])[0]

# Cruce entre individuos
def cruce(padre1, padre2):
    punto_corte = random.randint(1, 3)
    nuevo_operadores = padre1[1][:punto_corte] + padre2[1][punto_corte:]
    nuevo_numeros = list(set(padre1[0][:3] + padre2[0][2:]))[:5]
    return nuevo_numeros, nuevo_operadores

# Mutación de individuos
def mutacion(individuo):
    if random.random() < 0.3:
        idx_op = random.randint(0, len(individuo[1]) - 1)
        idx_num = random.randint(0, len(individuo[0]) - 1)
        individuo[1][idx_op] = random.choice(operadores_disponibles)
        nuevo_numero = random.choice(numeros_disponibles)
        while nuevo_numero in individuo[0]:
            nuevo_numero = random.choice(numeros_disponibles)
        individuo[0][idx_num] = nuevo_numero
    return individuo

# Algoritmo Genético Principal
def algoritmo_genetico(iteraciones=300, tamano_poblacion=100):
    poblacion = generar_poblacion(tamano_poblacion)

    for _ in range(iteraciones):
        fitness_scores = [calcular_fitness(ind) for ind in poblacion]
        nueva_poblacion = []
        for _ in range(tamano_poblacion // 2):
            padre1 = seleccion(poblacion, fitness_scores)
            padre2 = seleccion(poblacion, fitness_scores)
            hijo1 = cruce(padre1, padre2)
            hijo2 = cruce(padre2, padre1)
            nueva_poblacion.extend([mutacion(hijo1), mutacion(hijo2)])
        poblacion = nueva_poblacion

    mejor_expresion, mejor_resultado = None, None
    valor_minimo, valor_maximo = float('inf'), float('-inf')
    expresion_minima, expresion_maxima = None, None

    for individuo in poblacion:
        ecuacion = construir_ecuacion(individuo[0], individuo[1])
        resultado = evaluar_ecuacion(ecuacion)

        if resultado is not None:
            if resultado == 0 and (mejor_resultado is None or abs(resultado) < abs(mejor_resultado)):
                mejor_expresion = ecuacion
                mejor_resultado = resultado
            if resultado < valor_minimo:
                valor_minimo = resultado
                expresion_minima = ecuacion
            if resultado > valor_maximo:
                valor_maximo = resultado
                expresion_maxima = ecuacion

    return mejor_expresion, mejor_resultado, valor_minimo, expresion_minima, valor_maximo, expresion_maxima

# Generar datos de entrada aleatorios
datos_entrada_aleatorios = generar_datos_entrada_aleatorios(10)

# Aplicar el algoritmo genético a los datos generados
resultados_algoritmo_genetico = []
for numeros, operadores, ecuacion, resultado in datos_entrada_aleatorios:
    mejor_expresion, mejor_resultado, valor_minimo, expresion_minima, valor_maximo, expresion_maxima = algoritmo_genetico()
    resultados_algoritmo_genetico.append((ecuacion, resultado, mejor_expresion, mejor_resultado, valor_minimo, expresion_minima, valor_maximo, expresion_maxima))

# Convertir los resultados en un DataFrame
df_resultados_genetico = pd.DataFrame(resultados_algoritmo_genetico, columns=[
    "Ecuación Original", "Resultado Original", "Mejor Expresión", "Mejor Resultado",
    "Valor Mínimo", "Expresión Mínima", "Valor Máximo", "Expresión Máxima"
])

# Mostrar resultados
print(df_resultados_genetico)


  Ecuación Original  Resultado Original Mejor Expresión  Mejor Resultado  \
0         9*6*3+4/2               164.0       1*2-3-8+9                0   
1         7*8+4/1+3                63.0       1*2-3-4+5                0   
2         7*2-3*5+1                 0.0       2+4-5+7-8                0   
3         1+5/7-6/2                 NaN       1/2*3*6-9                0   
4         4+3*6*8+2               150.0       1*2-3-4+5                0   
5         8/6/5/3*2                 NaN       9-4-6+1*1                0   
6         8-2*1/7+4                 NaN       1-4+5+6-8                0   
7         6-4*3*9*7              -750.0       2+3-6+7/7                0   
8         2/9/8/1*4                 NaN       1*3-4-7+8                0   
9         1+7/5-8-4                 NaN       1+3-5-7+8                0   

   Valor Mínimo Expresión Mínima  Valor Máximo Expresión Máxima  
0           -51        1-2*8-4*9            24        1+6*3-4+9  
1           -24        1*2-6-4*

# Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Referencias bibliográficas:

* Documentación oficial de Python (uso de random, eval, estructuras de datos)

🔗 https://docs.python.org/3/

* Algoritmos genéticos en Python – Artículo de GeeksforGeeks

🔗 https://www.geeksforgeeks.org/genetic-algorithms/

* Optimización con Algoritmos Genéticos – Towards Data Science

🔗 https://towardsdatascience.com/genetic-algorithm-explained-a-beginners-guide-to-nature-inspired-optimization-5055680219a7

* Enfriamiento simulado y optimización combinatoria – Coursera

🔗 https://www.coursera.org/lecture/metaheuristicas/algoritmo-de-enfriamiento-simulado-vqDlJ




Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta

El estudio de este problema abre un fascinante abanico de posibilidades en el ámbito de la optimización combinatoria y la inteligencia artificial aplicada. Inicialmente, el enfoque se centra en explorar todas las combinaciones posibles de cifras y operadores bajo las restricciones establecidas, utilizando métodos de fuerza bruta y algoritmos heurísticos como algoritmos genéticos para encontrar soluciones de manera eficiente. Sin embargo, a medida que el tamaño del problema escala, se pueden implementar técnicas avanzadas de búsqueda en espacio de estados, como programación dinámica o incluso el uso de redes neuronales para predecir estructuras matemáticas con mayor probabilidad de generar resultados enteros dentro de un rango determinado.

Las variaciones del problema pueden incluir la introducción de más operadores matemáticos avanzados, el uso de dígitos repetidos, o incluso la restricción de ciertas operaciones como la división para garantizar siempre números enteros. A nivel computacional, aumentar el tamaño del problema podría derivar en su estudio desde una perspectiva de computación cuántica, donde los algoritmos cuánticos de optimización podrían abordar la explosión combinatoria con mayor eficiencia. En última instancia, este problema no es solo un ejercicio matemático; es una puerta de entrada a la exploración de estrategias de búsqueda óptima, con aplicaciones en criptografía, análisis simbólico y la automatización del razonamiento matemático